sympy/solvers/decompogen.py

```diff
@@ -48,8 +48,26 @@ def decompogen(f, symbol):
             arg = f.args[0]
         if arg == symbol:
             return [f]
-        result += [f.subs(arg, symbol)] + decompogen(arg, symbol)
-        return result
+        return [f.subs(arg, symbol)] + decompogen(arg, symbol)
+
+    # ===== Min/Max Functions ===== #
+    if isinstance(f, (Min, Max)):
+        args = list(f.args)
+        d0 = None
+        for i, a in enumerate(args):
+            if not a.has_free(symbol):
+                continue
+            d = decompogen(a, symbol)
+            if len(d) == 1:
+                d = [symbol] + d
+            if d0 is None:
+                d0 = d[1:]
+            elif d[1:] != d0:
+                raise TypeError('cannot decompose %s' % f)
```

Comment:
Should this raise? It seems like it should just return `[f]`. Are there other places where it raises?

Reply:
> Should this raise? 

Yes, `decompogen(Max(sin(x), x**2), x))` raises error at that point and it should.

> Are there other places where it raises?

I'm not sure of this but otherwise we can add this check ,
```
if all(a.has(symbol) for a in f.args):
    raise TypeError('cannot decompose %s' % f)
```
at the beginning and return `[f]` here.

Reply:
I don't think it should raise. If you can't decompose it, it is in primitive form.
```python
>>> decompose(2*x+3)
[2*x + 3]
>>> decompogen(2*x+3,x)
[2*x + 3]
```
And as the added tests in the diff show, there is no reason to raise if all of the args don't contain `symbol`.

Reply:
I'm referring to this https://github.com/sympy/sympy/pull/18517#discussion_r373071166 where @sylee957 mentioned decompogen should not be able to decompose or compose functions with arguments like sin(x), x**2.

Reply:
I'm not sure if these need to be done, however. If it is just a linear decomposition then there is no computational advantage that I can think of.
```
                # TODO additive generators could be recognized
                # like Max(2*x + 1, (2*x + 3)**2) -> [Max(x, (x + 2)**2), 2*x + 1]
                # multiplicative could be recognized, too
                # like Max(2*x, 16*x**2) -> [Max(x, 4*x**2), 2*x]
```

Reply:
New diff
```diff
diff --git a/sympy/solvers/decompogen.py b/sympy/solvers/decompogen.py
index 51d380b86a..f64d592bcc 100644
--- a/sympy/solvers/decompogen.py
+++ b/sympy/solvers/decompogen.py
@@ -40,8 +40,6 @@ def decompogen(f, symbol):
     if symbol not in f.free_symbols:
         return [f]
 
-    result = []
-
     # ===== Simple Functions ===== #
     if isinstance(f, (Function, Pow)):
         if f.is_Pow and f.base == S.Exp1:
@@ -50,21 +48,29 @@ def decompogen(f, symbol):
             arg = f.args[0]
         if arg == symbol:
             return [f]
-        result += [f.subs(arg, symbol)] + decompogen(arg, symbol)
-        return result
+        return [f.subs(arg, symbol)] + decompogen(arg, symbol)
 
     # ===== Min/Max Functions ===== #
     if isinstance(f, (Min, Max)):
         args = list(f.args)
-        iargs = [i for i, a in enumerate(f.args) if a.has(symbol)]
-        if len(iargs) == len(args):
-            raise TypeError('cannot decompose %s' % f)
-        dec = []
-        for i in iargs:
-            dec.extend(decompogen(args[i], symbol))
-            args[i] = symbol
-        result += [f.func(*args)] + list(uniq(dec))
-        return result
+        d0 = None
+        for i, a in enumerate(args):
+            if not a.has_free(symbol):
+                continue
+            d = decompogen(a, symbol)
+            if len(d) == 1:
+                d = [symbol] + d
+            if d0 is None:
+                d0 = d[1:]
+            elif d[1:] != d0:
+                # decomposition is not the same for each arg:
+                # mark as having no decomposition
+                d0 = [symbol]
+                break
+            args[i] = d[0]
+        if d0 == [symbol]:
+            return [f]
+        return [f.func(*args)] + d0
 
     # ===== Convert to Polynomial ===== #
     fp = Poly(f)
@@ -73,13 +79,11 @@ def decompogen(f, symbol):
     if len(gens) == 1 and gens[0] != symbol:
         f1 = f.subs(gens[0], symbol)
         f2 = gens[0]
-        result += [f1] + decompogen(f2, symbol)
-        return result
+        return [f1] + decompogen(f2, symbol)
 
     # ===== Polynomial decompose() ====== #
     try:
-        result += decompose(f)
-        return result
+        return decompose(f)
     except ValueError:
         return [f]
 
diff --git a/sympy/solvers/tests/test_decompogen.py b/sympy/solvers/tests/test_decompogen.py
index 22bcde1d44..e616b322af 100644
--- a/sympy/solvers/tests/test_decompogen.py
+++ b/sympy/solvers/tests/test_decompogen.py
@@ -19,11 +19,11 @@ def test_decompogen():
     assert decompogen(Abs(cos(y)**2 + 3*cos(x) - 4), x) == [Abs(x), 3*x + cos(y)**2 - 4, cos(x)]
     assert decompogen(x, y) == [x]
     assert decompogen(1, x) == [1]
-    assert decompogen(Max(3, x), x) == [Max(3, x), x]
+    assert decompogen(Max(3, x), x) == [Max(3, x)]
     raises(TypeError, lambda: decompogen(x < 5, x))
     u = 2*x + 3
-    raises(TypeError, lambda: decompogen(Max(u, u**2), x))
-    assert decompogen(Max(u, u**2, y), x) == [Max(x, y), 2*x + 3, x**2]
+    assert decompogen(Max(sqrt(u),(u)**2), x) == [Max(sqrt(x), x**2), u]
+    assert decompogen(Max(u, u**2, y), x) == [Max(x, x**2, y), u]
 
 
 def test_decompogen_poly():
```

Reply:
> I'm not sure if these need to be done

Maybe they can also be done.

> ```
>                 # TODO additive generators could be recognized
>                 # like Max(2*x + 1, (2*x + 3)**2) -> [Max(x, (x + 2)**2), 2*x + 1]
>                 # multiplicative could be recognized, too
>                 # like Max(2*x, 16*x**2) -> [Max(x, 4*x**2), 2*x]
> ```

Should I also add these `TODO` lines to the code? If we are not implementing these changes in this PR.

Reply:
I'm not sure it's worth it. I suspect that Min/Max will often just come back as `[f]`.